flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphDeploymentTest.java

```diff
@@ -578,21 +576,23 @@ public void testExecutionGraphIsDeployedInTopologicalOrder() throws Exception {
 			return CompletableFuture.completedFuture(Acknowledge.get());
 		});
 
+		final TaskManagerLocation taskManagerLocation = new LocalTaskManagerLocation();
 		final TestingTaskExecutorGateway taskExecutorGateway = testingTaskExecutorGatewayBuilder.createTestingTaskExecutorGateway();
 		final RpcTaskManagerGateway taskManagerGateway = new RpcTaskManagerGateway(taskExecutorGateway, JobMasterId.generate());
 
-		final Collection<CompletableFuture<LogicalSlot>> slotFutures = new ArrayList<>(numberTasks);
-		for (int i = 0; i < numberTasks; i++) {
-			slotFutures.add(new CompletableFuture<>());
-		}
-
-		final SlotProvider slotProvider = new IteratorTestingSlotProvider(slotFutures.iterator());
+		final Collection<CompletableFuture<TestingPhysicalSlot>> slotFutures = new ArrayList<>(numberTasks);
 
 		final JobGraph jobGraph = new JobGraph(jobId, "Test Job", sourceVertex, sinkVertex);
 		jobGraph.setScheduleMode(ScheduleMode.EAGER);
 
-		final SchedulerBase scheduler = SchedulerTestingUtils
-			.newSchedulerBuilderWithDefaultSlotAllocator(jobGraph, slotProvider)
+		final SchedulerBase scheduler = SchedulerTestingUtils.newSchedulerBuilder(jobGraph)
+			.setExecutionSlotAllocatorFactory(SchedulerTestingUtils.slotSharingExecutionSlotAllocatorFactoryBuilder()
+				.withSlotProvider(TestingPhysicalSlotProvider.builder()
+					.withPhysicalSlotCreator((f, a, tml, tmg, r) -> slotFutures.add(f))
```

Comment:
Ah ok, now I see why you have introduced the consumer. I am not 100% sold that this is the best way to achieve it. Maybe we can collect all the `SlotRequestIds` from the `TestingPhysicalSlotProvider` and then call `TestingPhyiscalSlotProvider.completePhysicalSlotFutureFor` in a random order?